src/utils/pluginDriver.ts

```diff
@@ -0,0 +1,194 @@
+import Graph from '../Graph';
+import {
+	InputOptions,
+	Plugin,
+	PluginContext,
+	RollupError,
+	RollupWarning,
+	Watcher
+} from '../rollup/types';
+import { createAssetPluginHooks, EmitAsset } from './assetHooks';
+import { getRollupDefaultPlugin } from './default-plugin';
+import error from './error';
+
+export interface PluginDriver {
+	emitAsset: EmitAsset;
+	getAssetFileName(assetId: string): string;
+	hookSeq(hook: string, args?: any[], context?: HookContext): Promise<void>;
+	hookFirst<T = any>(hook: string, args?: any[], hookContext?: HookContext): Promise<T>;
+	hookParallel(hook: string, args?: any[], hookContext?: HookContext): Promise<void>;
+	hookReduceArg0<R = any, T = any>(
+		hook: string,
+		args: any[],
+		reduce: Reduce<R, T>,
+		hookContext?: HookContext
+	): Promise<T>;
+	hookReduceValue<R = any, T = any>(
+		hook: string,
+		value: T | Promise<T>,
+		args: any[],
+		reduce: Reduce<R, T>,
+		hookContext?: HookContext
+	): Promise<T>;
+	hasLoadersOrTransforms: boolean;
+}
+
+export type Reduce<R = any, T = any> = (reduction: T, result: R, plugin: Plugin) => T;
+export type HookContext = (context: PluginContext, plugin?: Plugin) => PluginContext;
+
+export function createPluginDriver(
+	graph: Graph,
+	options: InputOptions,
+	watcher?: Watcher
+): PluginDriver {
+	const plugins = [...(options.plugins || []), getRollupDefaultPlugin(options)];
+	const { emitAsset, getAssetFileName, setAssetSource } = createAssetPluginHooks(graph.assetsById);
+
+	let hasLoadersOrTransforms = false;
+
+	const pluginContexts = plugins.map(plugin => {
+		if (
+			!hasLoadersOrTransforms &&
+			(plugin.load || plugin.transform || plugin.transformBundle || plugin.transformChunk)
+		)
+			hasLoadersOrTransforms = true;
+
+		const context: PluginContext = {
+			watcher,
+			isExternal(id: string, parentId: string, isResolved = false) {
+				return graph.isExternal(id, parentId, isResolved);
+			},
+			resolveId(id: string, parent: string) {
+				return pluginDriver.hookFirst('resolveId', [id, parent]);
+			},
+			parse: graph.contextParse,
+			emitAsset,
+			getAssetFileName,
+			setAssetSource,
+
+			warn: (warning: RollupWarning | string) => {
+				if (typeof warning === 'string') warning = { message: warning };
+				if (warning.code) warning.pluginCode = warning.code;
+				warning.code = 'PLUGIN_WARNING';
+				warning.plugin = plugin.name || '(anonymous plugin)';
+				graph.warn(warning);
+			},
+			error: (err: RollupError | string) => {
+				if (typeof err === 'string') err = { message: err };
+				if (err.code) err.pluginCode = err.code;
+				err.code = 'PLUGIN_ERROR';
+				err.plugin = plugin.name || '(anonymous plugin)';
+				error(err);
```

Comment:
Maybe the code to normalize warnings and errors should actually be extracted into a shared function so that this can become something like
`error(normalizePluginError('PLUGIN_ERROR'))` and `graph.warn(normalizePluginError('PLUGIN_WARNING'))?

Reply:
Agreed this should be a next step handling error centralization. Would prefer to do that in successive PR work.

Reply:
👍